src/index.ts

```diff
@@ -39,20 +39,15 @@ function addMethod(schemaType: any, name: string, fn: any) {
 
 type ObjectSchemaOf<T extends AnyObject> = ObjectSchema<
   {
-    [k in keyof T]-?: T[k] extends Array<infer E>
-      ? ArraySchema<SchemaOf<E>>
-      : T[k] extends AnyObject
-      ? // we can't use  ObjectSchema<{ []: SchemaOf<T[k]> }> b/c TS produces a union of two schema
-        ObjectSchemaOf<T[k]>
-      : BaseSchema<Maybe<T[k]>, AnyObject, T[k]>;
+    [k in keyof T]-?: SchemaOf<T[k]> | Lazy<SchemaOf<T[k]>>;
```

Comment:
see the old comment here. We can't use `SchemaOf<T[k]>` because of how TS applies the types over the conditional in SchemaOf. Basically if the input type is something like `string | undefined`, you get `SchemaOf<string> | SchemaOf<undefined>`, not `SchemaOf<string | undefined>`, The behavior is different for passing the generic to a concrete type like `BaseSchema`. Just means some duplication

Reply:
Okay... I'm really confused about what's going on here. Either I'm doing something wrong, or it seems like that comment is backwards. Can you look at the minimal example I made in this commit? https://github.com/jquense/yup/pull/1170/commits/8bf7689021379cc2ad47f3c34f705dd556b8889f

Reply:
I somewhat suspect the issue is with my example code (`FormData`/`buildNestedValidator`), not the Yup type definitions... but I can't figure out the right way, either. I'm not sure how to translate my recursive `FormData` to a corresponding `SchemaOf` type.

Reply:
Ah ha, I think I figured out what's going on. I'll get back to you soon with this sorted out. Recursive types are... fun.

Reply:
Okay, I added a commit that adds lazy to field values for object schemas.

And in case you wanted an explanation on how I got so confused:
```typescript
interface FormData {
  [K: string]: string | FormData;
}
type FormDataSchema = ObjectSchema<{
  [K: string]: SchemaOf<string> | FormDataSchema;
}>;
export const buildNestedValidator = (
  path: string[],
  validator: SchemaOf<string> | FormDataSchema,
): FormDataSchema => {
  const [head, ...rest] = path;
  const result: SchemaOf<string> | FormDataSchema =
    rest.length === 0 ? validator : buildNestedValidator(rest, validator);
  return objectCreate({ [head]: result });
};
const foo: FormData = { foo: '', bar: { baz: '' } };
const val = buildNestedValidator(['bar', 'baz'], stringCreate().required());
val.validate(foo);
```

I was trying to use `SchemaOf<FormData>` originally, but that wasn't working. The solution was to create the `FormDataSchema` type and use it instead. I think that's because of how the union gets distributed - e.g. `SchemaOf<string | string[]>` vs `SchemaOf<string> | SchemaOf<string[]`. Something along those lines - my brain hurts at the moment.

Anyways, that's all irrelevant to the actual change in this PR and is just context on how I got so tripped up testing against my own code base. I went ahead and rebased to squash my mess - **this is ready for you to review and potentially merge now**.